# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

1. Identify the dimensions of the dataset by determining the number of rows and columns it contains.

In [1]:
import numpy as np 
import pandas as pd

In [2]:
customer_data = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv')

In [3]:
display(customer_data.head())
display(customer_data.tail())

Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

Customer   ST GENDER Education Customer Lifetime Value  Income  \
4003      NaN  NaN    NaN       NaN                     NaN     NaN   
4004      NaN  NaN    NaN       NaN                     NaN     NaN   
4005      NaN  NaN    NaN       NaN                     NaN     NaN   
4006      NaN  NaN    NaN       NaN                     NaN     NaN   
4007      NaN  NaN    NaN       NaN                     NaN     NaN   

      Monthly Premium Auto Number of Open Complaints Policy Type  \
4003                   NaN                       NaN         NaN   
4004                   NaN                       NaN         NaN   
4005                   NaN                       NaN         NaN   
4006                   NaN                       NaN         NaN   
4007                   NaN                       NaN         NaN   

     Vehicle Class  Total Claim Amount  
4003           NaN                 NaN  
4004           NaN                 NaN  
4005           NaN                 NaN  
4006           NaN                 NaN  
4007           NaN                 NaN

In [4]:
print(customer_data.shape)
print("The dataframe contains", customer_data.shape[0], "rows and", customer_data.shape[1], "columns.")

(4008, 11)
The dataframe contains 4008 rows and 11 columns.


2. Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.


In [5]:
customer_data.dtypes

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object

A. Customer: I would consider that the customer should have integer as data type because it's a numerical discrete variable. Nevertheless, it's understandable that in this case it's considered as object because the values have letters on it.

B. ST: the data type "object" is correct because the state names represent strings.

C. GENDER: the data type "object" is correct be"ause it's a categorical nominal variable.

D. Education: the data type "object" is correct because it's a categorical ordinal variable.

E. Customer Lifetime Value: if we consider that this is a numerical continuous variable, the data type "object" is not correct but, as it contains the percentage symbol on it, it's defined as a string. Fix suggestion: considering that this KPI shows revenue and not percentange, I would create a function to eliminate the "%" from the cells (in case the business agrees that the percentage symbol is incorrect).

F. Income: the data type "float64" is correct because this is a numerical continuous variable.

G. Monthly Premium Auto: the data type "float64" is correct because this is a numerical continuous variable.

H. Number of Open Complaints: the data type "object" is incorrect if we consider that this variable is numerical discrete (it should be an integer). From my point of view, before performing any modification we need to understand what each number between the slashes mean. *** View comments on exercise 3.

I. Policy Type: the data type "object" is correct because it's a categorical ordinal variable.

J. Vehicle Class: the data type "object" is correct because it's a categorical ordinal variable.

K. Total Claim Amount: the data type "float64" is correct because this is a numerical continuous variable.

3. Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.

In [6]:
customer_data.nunique()

Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64

If we take a look at the unique values, ST, GENDER, Educations, Policy Type, Vehicle Class and Number of Open Complaints would be categorical. Nevertheless, I would like to have more information about this last column and their values before manipulating the data.

In [7]:
customer_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4008 entries, 0 to 4007
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   1071 non-null   object 
 1   ST                         1071 non-null   object 
 2   GENDER                     954 non-null    object 
 3   Education                  1071 non-null   object 
 4   Customer Lifetime Value    1068 non-null   object 
 5   Income                     1071 non-null   float64
 6   Monthly Premium Auto       1071 non-null   float64
 7   Number of Open Complaints  1071 non-null   object 
 8   Policy Type                1071 non-null   object 
 9   Vehicle Class              1071 non-null   object 
 10  Total Claim Amount         1071 non-null   float64
dtypes: float64(3), object(8)
memory usage: 344.6+ KB


In [8]:
customer_data["ST"].value_counts()

ST
Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64

Insights: if we consider the available data (with no missing values), most of the customers live in the state of California (331 after data cleaning) while the state with less customers is Nevada (98). The sum of all state's unique values is 1071 and the unique amount of customer is also 1071, therefore all customers have one state associated to them.

In [9]:
customer_data["GENDER"].value_counts()

GENDER
F         457
M         413
Male       39
female     28
Femal      17
Name: count, dtype: int64

Insights: after reviewing the two previous commands, we can conclude that some customers don't have data related to their gender. We have 1071 customers but only 954 values filled related to their gender. Therefore, around 11% of the data (117 values) is missing. If considering only the available data, the distribution between genres is very similar: 53% are women and 47% are men.

In [10]:
customer_data["Education"].value_counts()

Education
Bachelor                324
College                 313
High School or Below    296
Master                   94
Doctor                   37
Bachelors                 7
Name: count, dtype: int64

Insights: all customers have their education level associated to them. Most of them have a bachelor degree (331) followed by 313 with a college degree. Only 37 customers have a doctor's degree.

In [11]:
customer_data["Number of Open Complaints"].value_counts()

Number of Open Complaints
1/0/00    830
1/1/00    138
1/2/00     50
1/3/00     34
1/4/00     13
1/5/00      6
Name: count, dtype: int64

Insights: all customers have data related to number of complaints associated to them. Most of the customers don't have any complaints (830 - ca. 77% of data). Only 6 customers have the maximum amount of complaints (5).

In [12]:
customer_data["Policy Type"].value_counts()

Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64

Insights: all customers have policy type data associated to them. Most of them (ca. 73%) have a Personal Auto policy while only 5% of them have a Special Auto one.

In [13]:
customer_data["Vehicle Class"].value_counts()

Vehicle Class
Four-Door Car    576
Two-Door Car     205
SUV              199
Sports Car        57
Luxury SUV        20
Luxury Car        14
Name: count, dtype: int64

Insights: all customers have vehicle class data associated to them. Most of them have a four-door car (576) while only 34 of them have a luxury car.

In [14]:
customer_data['Income'].min(), customer_data['Income'].max()

(0.0, 99960.0)

Insights: the biggest income across customers is 99960.0 while the minimum is 0.0 (we would need to ask the business which are the considerations to have in mind when filling this data).

In [15]:
customer_data['Monthly Premium Auto'].min(), customer_data['Monthly Premium Auto'].max()

(61.0, 35354.0)

Insights: the minimum amount spent per month is 61.0 and the biggest is 35354.0

In [16]:
customer_data['Total Claim Amount'].min(), customer_data['Total Claim Amount'].max()

(0.382107, 2893.239678)

Insights: the minimum value of approved claim amounts is 0.38 and the maximum is 2893.24

4. Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.

In [17]:
customer_data_no_na = customer_data.dropna(how='all') # Dropping all columns that only have missing values
display(customer_data_no_na)

Customer          ST GENDER             Education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
1066  TM65736      Oregon      M                Master   
1067  VJ51327        Cali      F  High School or Below   
1068  GS98873     Arizona      F              Bachelor   
1069  CW49887  California      F                Master   
1070  MY31220  California      F               College   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
1066              305955.03%  38644.0                  78.0   
1067             2031499.76%  63209.0                 102.0   
1068              323912.47%  16061.0                  88.0   
1069              462680.11%  79487.0                 114.0   
1070              899704.02%  54230.0                 112.0   

     Number of Open Complaints     Policy Type  Vehicle Class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
1066                    1/1/00   Personal Auto  Four-Door Car   
1067                    1/2/00   Personal Auto            SUV   
1068                    1/0/00   Personal Auto  Four-Door Car   
1069                    1/0/00    Special Auto            SUV   
1070                    1/0/00   Personal Auto   Two-Door Car   

      Total Claim Amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
1066          361.455219  
1067          207.320041  
1068          633.600000  
1069          547.200000  
1070          537.600000  

[1071 rows x 11 columns]

In [18]:
customer_data_no_na.describe() # As I don't know how the NaN rows affect the statistical values, I decided to get the describe() from the new dataframe with no NaN rows.

Income  Monthly Premium Auto  Total Claim Amount
count   1071.000000           1071.000000         1071.000000
mean   39295.701214            193.234360          404.986909
std    30469.427060           1601.190369          293.027260
min        0.000000             61.000000            0.382107
25%    14072.000000             68.000000          202.157702
50%    36234.000000             83.000000          354.729129
75%    64631.000000            109.500000          532.800000
max    99960.000000          35354.000000         2893.239678

In [19]:
customer_data_no_na['Income'].median(), customer_data_no_na['Monthly Premium Auto'].median(), customer_data_no_na['Total Claim Amount'].median()

(36234.0, 83.0, 354.729129)

In [20]:
customer_data_no_na['Income'].mode(), customer_data_no_na['Monthly Premium Auto'].mode(), customer_data_no_na['Total Claim Amount'].mode()

(0    0.0
 Name: Income, dtype: float64,
 0    65.0
 Name: Monthly Premium Auto, dtype: float64,
 0    321.6
 Name: Total Claim Amount, dtype: float64)

Insights:
1. The average income among customer is ca. 36k. Customers spend ca. 83.0 as premium in a monthly basis (we consider the median as, in this case, the mean is affected a lot by extreme values - 193.23). Also, 354.8 is the average of total claim amount among customers. Having said that, 50% of the customers are considered within these values.

2. Most customers pay 65.0 as premium in a monthly basis while most common total claim amount is 321.6. The most repeated income amount is 0.0 but we need more information from the business in order to understand how this data is inputted. After that, we will be able to provide more insights about this.

3. 68% of the customers have an income between 30469k and 48121k and a total claim amount between 293.03 and 698.02. --- Monthly Premium Auto: as the mean is hughly affected by extreme values, I don't really know how to analize the standard deviation on this case. Should I consider the median in this case?

4. 25% of customers have less than 14k of income and less than 202.16 of claim amount. 75% of the customers (in which the previous 25% is included) have an income of less than 64.6k and less than 532.8 of claim amount.

4. Insights on max and min amounts were provided in the previous exercise.

5. Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

Insights: summary statistics for these variables were defined in exercise 3.

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [27]:
# Create a nested list considering the value counts data from the column "ST"

customer_location = [['Oregon', 320],
                      ['California', 211],
                      ['Arizona', 186],
                      ['Cali', 120],
                      ['Nevada', 98],
                      ['Washington', 81],
                      ['WA', 30],
                      ['AZ', 25]]

# Create a Dataframe object

df_customer_location = pd.DataFrame(customer_location,
                  columns=['Location','Frequency'],
                  index=[0,1,2,3,4,5,6,7])

# Replace values such as Cali, WA and AZ for the correct ones

df_customer_location['Location'].replace('Cali', 'California', inplace=True)
df_customer_location['Location'].replace('WA', 'Washington', inplace=True)
df_customer_location['Location'].replace('AZ', 'Arizona', inplace=True)

display(df_customer_location)

Location  Frequency
0      Oregon        320
1  California        211
2     Arizona        186
3  California        120
4      Nevada         98
5  Washington         81
6  Washington         30
7     Arizona         25

In [28]:
# Sum the frequencies of locations with the same values

df_customer_location_sum = df_customer_location.groupby('Location', as_index=False)['Frequency'].sum()
display(df_customer_location_sum)

Location  Frequency
0     Arizona        211
1  California        331
2      Nevada         98
3      Oregon        320
4  Washington        111

In [31]:
# Sort values

df_sorted = df_customer_location_sum.sort_values("Frequency")
df_sorted2 = df_sorted.reset_index(drop=True)
display(df_sorted2)

Location  Frequency
0      Nevada         98
1  Washington        111
2     Arizona        211
3      Oregon        320
4  California        331

### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [39]:
# Create a nested list considering the value counts data from the column "Policy Type"

policy_type = [['Personal Auto', 780],
               ['Corporate Auto', 234],
               ['Special Auto', 57]]

# Create a Dataframe object

df_policy_type = pd.DataFrame(policy_type,
                              columns=['Policy type', 'Sold'],
                              index=[0,1,2])

display(df_policy_type)

Policy type  Sold
0   Personal Auto   780
1  Corporate Auto   234
2    Special Auto    57

In [48]:
# Use loc and idxmax() method to bring policy type with the highest number of policies sold

df_policy_type.loc[df_policy_type['Sold'].idxmax()]

Policy type    Personal Auto
Sold                     780
Name: 0, dtype: object

### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [49]:
# Create general dataframe with policy and income data

df_policy_income = customer_data_no_na.loc[:,["Policy Type","Income"]]
display(df_policy_income)

Policy Type   Income
0      Personal Auto      0.0
1      Personal Auto      0.0
2      Personal Auto  48767.0
3     Corporate Auto      0.0
4      Personal Auto  36357.0
...              ...      ...
1066   Personal Auto  38644.0
1067   Personal Auto  63209.0
1068   Personal Auto  16061.0
1069    Special Auto  79487.0
1070   Personal Auto  54230.0

[1071 rows x 2 columns]

In [94]:
# Create dataframe for Personal Auto policy type and calculate average

df_personal_auto_income = df_policy_income[df_policy_income['Policy Type'] == "Personal Auto"]
print("The average income of customers with a Personal Auto policy is", df_personal_auto_income['Income'].mean())

The average income of customers with a Personal Auto policy is 38180.69871794872


In [95]:
# Create dataframe for Corporate Auto policy type and calculate average

df_corporate_auto_income = df_policy_income[df_policy_income['Policy Type'] == "Corporate Auto"]
print("The average income of customers with a Corporate Auto policy is", df_corporate_auto_income['Income'].mean())

The average income of customers with a Corporate Auto policy is 41390.31196581197


Answer: customers with a Personal Auto policy have a lower income than those with a Corporate Auto policy

### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [74]:
# Create dataframe only with Total Claim Amounts 

df_total_claim_amount = customer_data_no_na.loc[:,["Total Claim Amount"]]
display(df_total_claim_amount)

Total Claim Amount
0               2.704934
1            1131.464935
2             566.472247
3             529.881344
4              17.269323
...                  ...
1066          361.455219
1067          207.320041
1068          633.600000
1069          547.200000
1070          537.600000

[1071 rows x 1 columns]

In [92]:
# We already know - by previous exercises - that the quantile 75% is 532.8
# Filter for those amounts that are higher than the percentile 75%

quantile_value = df_total_claim_amount.quantile(q=0.75)
display(quantile_value)

df_total_claim_amount[df_total_claim_amount['Total Claim Amount'] >  532.800000]

Total Claim Amount    532.8
Name: 0.75, dtype: float64

Total Claim Amount
1            1131.464935
2             566.472247
17            647.442031
23            554.376763
26            799.200000
...                  ...
1059          720.752945
1061          546.524896
1068          633.600000
1069          547.200000
1070          537.600000

[264 rows x 1 columns]

In [93]:
df_total_claim_amount[df_total_claim_amount['Total Claim Amount'] >  532.800000].describe()

Total Claim Amount
count          264.000000
mean           782.228263
std            292.751640
min            537.600000
25%            606.521741
50%            679.597985
75%            851.400000
max           2893.239678